Force retry when adoption fails

stbenjam opened the pull request:
Currently when adoption fails, we use RegistrationError but this gets
cleared any time registration succeeds. We need a separate signal for
adoption failure, which allows us to force retry the adoption again.

Fixes #697

zaneb commented on apis/metal3.io/v1alpha1/baremetalhost_types.go:
```diff
@@ -123,6 +123,9 @@ const (
 const (
+	// AdoptionError is an error condition occurring when the controller
+	// is unable to adopt an already provisioned host.
+	AdoptionError ErrorType = "adoption error"
```
Need to make this an allowed value on line 525 also, otherwise we won't be able to write the status.

stbenjam replied:
Thanks! That was the trick - it shows up in the status field now

zaneb commented on controllers/metal3.io/host_state_machine.go:
```diff
@@ -183,6 +183,7 @@ func (hsm *hostStateMachine) ensureRegistered(info *reconcileInfo) (result actio
 	case metal3v1alpha1.StateRegistering:
 	default:
 		needsReregister = (hsm.Host.Status.ErrorType == metal3v1alpha1.RegistrationError ||
+			hsm.Host.Status.ErrorType == metal3v1alpha1.AdoptionError ||
```
I've convinced myself this is not required. If it fails in adoption then registration would have succeeded just prior, so we don't need to report that we're starting a new registration.

stbenjam replied:
Yea I wasn't sure but after testing you're right, we don't need this.

zaneb commented:
I've been thinking about alternative names that could avoid exposing to the user the implementation detail that after re-registration in the provisioned state, we call Ironic's adopt API.
The best I've come up with is "provisioned registration error" which is admittedly terrible. Any other ideas?

andfasano commented on controllers/metal3.io/baremetalhost_controller.go:
```diff
@@ -463,7 +464,10 @@ func (r *BareMetalHostReconciler) actionRegistering(prov provisioner.Provisioner
+
+	if info.host.Status.ErrorType == metal3v1alpha1.RegistrationError || registeredNewCreds {
+		clearError(info.host)
+	}
```
Suggestion: if we're going to fix the issue here you then if you want tests from #757 can be reused (and I'm going to update it accordingly just to fix the log traces)

stbenjam replied:
Ok makes sense - I pulled in your commit over without the logging changes, although the unit test is failing so I guess good call bringing the test over. Not sure what I broke, though.

andfasano replied:
I should have fixed them in #757 (due the latest rebase)

zaneb commented on controllers/metal3.io/host_state_machine_test.go:
```diff
@@ -289,6 +289,47 @@ func TestErrorCountCleared(t *testing.T) {
+			assert.Equal(t, tt.Host.Status.OperationalStatus, v1alpha1.OperationalStatusOK)
+			assert.Empty(t, tt.Host.Status.ErrorType)
+			assert.Empty(t, tt.Host.Status.ErrorMessage)
+			assert.Empty(t, tt.Host.Status.ErrorCount)
```
`clearError()` no longer resets the ErrorCount since #749 merged. That's why this test was passing earlier (in Andrea's PR) but failing now.
I think we can just get rid of this assertion.

stbenjam commented:
/test-integration

zaneb commented:
This LGTM.
@dhellmann do you have any opinion on the name of this error?

andfasano reviewed (approved):
It also LGTM

dhellmann commented:
> This LGTM.
> @dhellmann do you have any opinion on the name of this error?

The only time this error would come up is if the pod is rescheduled and we have to restore Ironic's state? That feels like we're exposing an internal error to the user. Can they even do anything about it?

I haven't followed the discussion, but is there a technical reason to not use `RegistrationError`? Maybe we have to be able to tell the 2 errors apart inside the BMO code to trigger the retry in the right way?

stbenjam commented:
> > This LGTM.
> > @dhellmann do you have any opinion on the name of this error?
> 
> The only time this error would come up is if the pod is rescheduled and we have to restore Ironic's state? That feels like we're exposing an internal error to the user. Can they even do anything about it?
> 
> I haven't followed the discussion, but is there a technical reason to not use `RegistrationError`? Maybe we have to be able to tell the 2 errors apart inside the BMO code to trigger the retry in the right way?

Yes.  Registration previously cleared out all error messages, I've refactored it to specifically look for RegistrationError instead, and introduced AdoptionError when there's a failure to adopt. Currently if adoption fails we never retry because the registration state clears all errors out.

dhellmann commented:
Do we have a corresponding state to indicate that a host is being adopted? I'm trying to understand if this new error condition exposes a concept to the user that we haven't previously expected them to know about. If not, then I might just call it something like `InternalError`. That's pretty vague, but doesn't leave the user wondering why a host is showing an error when they didn't do anything to it.

Since we don't really want to expose the error to the user, it would be even better if we could use ironic's status directly, but I don't think we use that pattern anywhere else.

zaneb commented:
Yeah, so it's kind of a philosophical question.
Adoption is not mentioned anywhere else in the public API, so this is a new and potentially confusing piece of information to users. (If 'adoption' means anything to users, it's probably the thing that the ExternallyProvisioned flag does.)

Ideally the fact that Ironic keeps a bunch of state in a relational database that occasionally gets deleted and needs to be regenerated should be something that users never need to know about. However, if something goes wrong with that part, then we need a way of recording it. Arguably we do already expose this to the user in the sense that a registration error can occur at any time, not just in the Registration state. But having an adoption error seems worse because of the complete meaninglessness of the name outside of the implementation detail that we use in Ironic to accomplish it. Nevertheless we must have some way of distinguishing errors in adoption from errors in registration to fix the bug.

My thought process behind `provisioned registration error` was basically that adoption is the second half of the registration process that only occurs when the Host is already provisioned. I think that's a productive line of enquiry even though that particular name is pretty so-so.

andfasano commented:
The idea of `InternalError` looks appealing, as it could be of some use in future - at the same time it risks to be abused being so generic, whereas the `provisioned registration error` is definitely more granular and strictly scoped for a very specific kind of error. Anyhow, by looking at the current Provisioner design, where many actions could depend on `ErrorType` (the `force` bool), I'd say that we're moving into a direction closer to having more granular errors (maybe in future we could introduce a `SubErrorType`?) rather than generic. 
The name `provisioned registration error` doesn't sound too bad to me (and I don't have better alternatives), so I think we could proceed in that direction.

stbenjam commented:
Thanks all, I've updated the message to the proposed name. Please take another look

/test-integration

zaneb commented:
/approve
Will leave this open until tomorrow in case somebody else has a better idea for the name.

furkatgofurov7 commented:
This also LGTM. 
I do not have either better alternative for the name, but `provisioned registration error` sounds reasonable.

furkatgofurov7 commented:
This needs a rebase though. 

/cc @stbenjam

stbenjam commented:
Thanks, rebased!


/test-integration

zaneb commented:
/lgtm
